### PR TITLE
RFC35: Common ancestor approach

### PR DESCRIPTION
**Pull request by manav2401**

### CODE REVIEW

File: eth/downloader/downloader.go [hunk 1]
```diff
@@ -764,12 +775,51 @@ func calculateRequestSpan(remoteHeight, localHeight uint64) (int64, int, int, ui
...
+func (d *Downloader) isValidChain(p *peerConnection, remoteHeader *types.Header) (bool, error) {
+	// We want to validate the chain by comparing the last checkpointed block
+	// we're storing in `checkpointWhitelist` with the peer's block.
+
+	// Check for availaibility of the last checkpointed block.
+	// This can be also be empty if our heimdall is not responsing
+	// or we're running without it.
+	if len(d.checkpointWhitelist) <= 0 {
+		// worst case, we don't have the checkpoints in memory
+		return true, nil
```
Comment (JekaMas) [inline eth/downloader/downloader.go:789]: shall we trust then to a remote peer?

Comment (manav2401) [reply eth/downloader/downloader.go:789]: I am not really sure. It's just that there can be multiple reasons for not having an entry in checkpoint whitelist map (node restart, heimdall down/syncing, running without heimdall, etc).

Comment (JekaMas) [reply eth/downloader/downloader.go:789]: I believe that by the end of the day we only trust to Heimdall node. Let's discuss such cases on the next reorg call.

File: consensus/bor/rest.go [hunk 2]
```diff
@@ -76,6 +89,21 @@ func (h *HeimdallClient) FetchStateSyncEvents(fromID uint64, to int64) ([]*Event
...
 
+// FetchLatestCheckpoint fetches the latest bor submitted checkpoint from heimdall
+func (h *HeimdallClient) FetchLatestCheckpoint() (*Checkpoint, error) {
+	var checkpoint Checkpoint
+
+	response, err := h.Fetch("/checkpoints/latest", "")
+	if err != nil {
+		return nil, err
+	}
+	if err := json.Unmarshal(response.Result, &checkpoint); err != nil {
```
Comment (JekaMas) [inline consensus/bor/rest.go:100]: you could use just `=` instead of `:=`

File: consensus/bor/rest.go [hunk 3]
```diff
@@ -76,6 +89,21 @@ func (h *HeimdallClient) FetchStateSyncEvents(fromID uint64, to int64) ([]*Event
 	return eventRecords, nil
 }
 
+// FetchLatestCheckpoint fetches the latest bor submitted checkpoint from heimdall
+func (h *HeimdallClient) FetchLatestCheckpoint() (*Checkpoint, error) {
+	var checkpoint Checkpoint
```
Comment (JekaMas) [inline consensus/bor/rest.go:94]: `checkpoint := &Checkpoint{}`
will simplify a bit

File: eth/downloader/downloader.go [hunk 4]
```diff
@@ -332,6 +339,10 @@ func (d *Downloader) LegacySync(id string, head common.Hash, td, ttd *big.Int, m
 	case nil, errBusy, errCanceled:
 		return err
 	}
+	if errors.Is(err, errCheckpointMismatch) {
+		// TODO: what better can be done here?
+		log.Warn("Mismatch in last checkpointed block", "peer", id, "err", err)
```
Comment (JekaMas) [inline eth/downloader/downloader.go:344]: shall we drop peer here?

File: eth/downloader/downloader.go [hunk 5]
```diff
@@ -1755,3 +1805,36 @@ func (d *Downloader) DeliverSnapPacket(peer *snap.Peer, packet snap.Packet) erro
 		return fmt.Errorf("unexpected snap packet type: %T", packet)
 	}
 }
+
+// Helper functions for checkpoint whitelist service
+
+// PurgeWhitelistMap purges data from checkpoint whitelist map
+func (d *Downloader) PurgeWhitelistMap() error {
+	for k := range d.checkpointWhitelist {
```
Comment (JekaMas) [inline eth/downloader/downloader.go:1813]: it might be faster to make a new map.

File: eth/handler_bor.go [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+)
+
+// fetchWhitelistCheckpoint fetched the latest checkpoint from it's local heimdall
+// and verifies the data against bor data.
+func (h *ethHandler) fetchWhitelistCheckpoint() (uint64, common.Hash, error) {
+	// check for checkpoint whitelisting: bor
+	checkpoint, err := h.chain.Engine().(*bor.Bor).HeimdallClient.FetchLatestCheckpoint()
+	if err != nil {
+		log.Debug("Failed to fetch latest checkpoint for whitelisting")
+		return 0, common.Hash{}, fmt.Errorf("failed to fetch latest checkpoint")
```
Comment (JekaMas) [inline eth/handler_bor.go:21]: could you create errors as global variables?

File: eth/downloader/downloader.go [hunk 7]
```diff
@@ -764,12 +775,51 @@ func calculateRequestSpan(remoteHeight, localHeight uint64) (int64, int, int, ui
 	return int64(from), count, span - 1, uint64(max)
 }
 
+// isValidChain checks if the chain we're about to receive from this peer is valid or not
```
Comment (JekaMas) [inline eth/downloader/downloader.go:778]: I love the way you comment code! :)

**Comment by manav2401** [comment]
Closing this in favour of https://github.com/maticnetwork/bor/pull/386

**Pull request closed without merging**